### PR TITLE
Include outward ticks in bounding box

### PR DESCRIPTION
**Pull request by mdboom**
Replaces #5502.

### CODE REVIEW

**Comment by mdboom** [comment]
This needs a test specifically of tight layout on outward ticks, but I thought I'd put this up for comment early anyway...

**Comment by QuLogic** [comment]
I think you meant this replaces #5502?

**Comment by mdboom** [comment]
> I think you meant this replaces #5502?

Oops.  You're right.

**Comment by QuLogic** [comment]
Well, I don't see anything particularly wrong with it, but are there no test images that use outward ticks? Or just none with a tight bounding box?

**Comment by mdboom** [comment]
A test has been added

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -567,7 +567,11 @@ def get_window_extent(self, *args, **kwargs):
         get the axes bounding box in display space; *args* and
         *kwargs* are empty
         """
-        return self.bbox
+        bbox = self.bbox
+        x_pad = self.xaxis.get_tick_padding()
+        y_pad = self.yaxis.get_tick_padding()
+        return mtransforms.Bbox([[bbox.x0 - x_pad, bbox.y0 - y_pad],
+                                 [bbox.x1 + x_pad, bbox.y1 + y_pad]])
```
Comment (WeatherGod) [inline lib/matplotlib/axes/_base.py]: Is this going to work for polar axes? I would also double-check that this doesn't utterly bork for axes3d (not that I'd expect it to work great there, just to not crash or something).

**Comment by WeatherGod** [comment]
image tests that have just been added are failing

**Comment by WeatherGod** [comment]
I don't think the test is really all that useful. tight_layout is for ensuring that subplots don't run into each other. So, a good test would be to have a grid of subplots, perhaps each with different tick directions?

File: lib/matplotlib/axes/_base.py (hunk 1 above)
Comment (mdboom) [inline lib/matplotlib/axes/_base.py]: polar axes don't have ticks -- but it does appear that this needs to be updated so it's aware of when the ticks are present or not.

It doesn't seem to crash for axes3d, though as you say, it doesn't really do the right thin other than add a bit more space.

**Comment by mdboom** [comment]
@WeatherGod: I've updated the test as you helpfully suggested.

File: lib/matplotlib/axis.py [hunk 2]
```diff
@@ -169,6 +169,21 @@ def apply_tickdir(self, tickdir):
...
 
+    def get_tickdir(self):
+        return self._tickdir
+
+    def get_tick_padding(self):
+        """
+        Get the length of the tick outside of the axes.
+        """
+        tickdir = self._tickdir
+        if tickdir == 'in':
```
Comment (tacaswell) [inline lib/matplotlib/axis.py]: bit of a bike shed, better to do this with a dictionary?

File: lib/matplotlib/axis.py [hunk 3]
```diff
@@ -1097,6 +1108,9 @@ def get_tightbbox(self, renderer):
         else:
             return None
 
+    def get_tick_padding(self):
+        return self.majorTicks[0].get_tick_padding()
```
Comment (tacaswell) [inline lib/matplotlib/axis.py]: should this be `max(self.majorTicks[0].get_tick_padding(), self.minorTicks[0].get_tick_padding()))` ? 

Never underestimate pathological users....

Comment (mdboom) [inline lib/matplotlib/axis.py]: That's a good point.

File: lib/matplotlib/axis.py (hunk 2 above)
Comment (mdboom) [inline lib/matplotlib/axis.py]: Sure.

File: lib/matplotlib/axis.py (hunk 3 above)
Comment (WeatherGod) [inline lib/matplotlib/axis.py]: Another point: are we always guaranteed at least one tick?

**Comment by mdboom** [comment]
I've addressed @tacaswell's comments, and added minor ticks to the test to catch the pathological case he mentions.

**Comment by mdboom** [comment]
Also updated to handle the case where there are no ticks, as suggested by @WeatherGod

**Comment by tacaswell** [comment]
backported as 2a6bb2653969a7c491add729af937d4d93d58b0f

**Comment by QuLogic** [comment]
Backport to v2.x was actually via 8ac5b4b28edf27ddebaea364bf8b3e685e709c9e.

**Pull request merged**